### DISCUSSION TITLE
useEffect / hook for windows in Next.js (window is not defined)

### DISCUSSION THREAD

**Discussion by mxcdh** [vercel/next.js#40684] [Help]
Hey

This is my hook:
```
export const useWindowSize = () => {
  const [windowSize, setWindowSize] = useState<number>(window.innerWidth);

  useEffect(() => {
    let lastWindowSize = window.innerWidth;

    const updateWindowSize = () => {
      setWindowSize(lastWindowSize);
    };
    window.addEventListener("resize", updateWindowSize); // add event listener
    return () => {
      window.removeEventListener("resize", updateWindowSize); // clean up
    };
  }, [windowSize]);

  return windowSize;
};
```
When i want use in page this hook, i get:
```
ReferenceError: window is not defined
```

**Comment by PsyGik** [comment]
Your initial state tries to access the window object, `useState<number>(window.innerWidth);`. This hook _may be_ running in your server side code, hence the error. Initalize it with `0`, or something else that suits your use case

**Comment by icyJoseph** [comment]
As said before, the problem here is that you are doing unguarded DOM/Window access, and when this code runs on the server, the absence of window makes it crash. A more resilient hook would be:

```ts

export const useWindowSize = () => {
    const [windowSize, setWindowSize] = useState(0);

    useEffect(() => {
        // inside useEffect, the window is always present
        const updateWindowSize = () => {
            setWindowSize(window.innerWidth);
        };

        updateWindowSize(); // as soon as we are on the client, run this handler

        window.addEventListener('resize', updateWindowSize); // works only on resize events

        return () => {
            window.removeEventListener('resize', updateWindowSize); // clean up
        };
    }, []); // attach this once

    return windowSize;
};
```

**Reply by Wachira2000** [reply]
@icyJoseph Kindly assist me here. Getting the same error 'window is not defined' @icyJoseph [https://prnt.sc/yssqFOel5rvD](url)

```
const Header = () => {
    const[active, setActive] = useState(false);

    useEffect(() => {
        const handleScroll = () => {
            if(typeof window !== 'undefined') {
            // detect scroll

            setActive(window.scrollY > 100);
        };

        // add event listener

        window.addEventListener('scroll', handleScroll);

        // clear event listener

        return () => {
            window.removeEventListener('scroll', handleScroll);
        };
}}, []);
```

**Reply by macnuoos** [reply]
Estou montando um projeto em reactjs, nextjs e utilizo three fiber e dri para criar objects 3d. Na ultima versão next 14.1.0 da o erro window not defined, relacionado troika-three-utils. Na versão anterior funciona normal, tentei corrigir mas não consegui. Pior que não acho nada referente ao erro na internet.  ref packges https://docs.pmnd.rs/react-three-fiber/getting-started/introduction and https://github.com/pmndrs/drei#readme. Com a versão 14.0.4 next funciona normal.

**Comment by Michelangelo-Foresta** [comment]
> ```ts
> export const useWindowSize = () => {
>     const [windowSize, setWindowSize] = useState(0);
> 
>     useEffect(() => {
>         // inside useEffect, the window is always present
>         const updateWindowSize = () => {
>             setWindowSize(window.innerWidth);
>         };
> 
>         updateWindowSize(); // as soon as we are on the client, run this handler
> 
>         window.addEventListener('resize', updateWindowSize); // works only on resize events
> 
>         return () => {
>             window.removeEventListener('resize', updateWindowSize); // clean up
>         };
>     }, []); // attach this once
> 
>     return windowSize;
> };
> ```
AGAIN. he said to initiliaze with 0. Works for me. Make sure you set state of windowSize inside useEffect()

**Comment by Wachira2000** [comment]
The issue was because the leaflet library was using SSR instead of client, resulting in the error 'window is not defined'. I solved this by importing dynamic from nextjs. Please see the below screenshot

https://prnt.sc/RWuE55T6skuA
